File: scripts/clean-scripts.ts
import { resolve } from 'path';

import dotenv from 'dotenv';
import glob from 'glob';
import { replaceInFile, ReplaceInFileConfig } from 'replace-in-file';

import logger from './logger';

dotenv.config();

const globSync = glob.sync;
const { AIRTABLE_BOOKMARKS_ID, AIRTABLE_TOKEN } = process.env;
const SITE_URL: string = 'https://cleverlaziness.com';
const timestamp: number = Math.floor(new Date().getTime() / 1000);

// Glob options. Pass directory to search and files to ignore
const cwd = resolve(__dirname, '..', 'dist');
const ignore = ['sw*.js', 'noise*.js', 'bookmarks*.js'];

// Find all JS, CSS, and font files in rendered output
(async () => {
  // create matched files array
  const files = globSync('**/*.{js,css,woff,woff2}', {
    cwd,
    ignore,
  });
  const sw = globSync('sw.min.*.js', { cwd });
  const noise = globSync('noise.min.*.js', { cwd });
  const noiseWW = globSync('noise.ww.min.*.js', { cwd });
  const newFiles = files.map(toCache => `'/${toCache}'`).toString();

  // find and replace options; add hash ID, files to cache array, and site base URL
  const replaceOptions: ReplaceInFileConfig = {
    files: [`${cwd}/${sw[0]}`, `${cwd}/${noise[0]}`],
    from: [/\["staticAssets"\]/g, /"version"/g, /baseURL/g, '/noise.ww.min.js'],
    to: [`[${newFiles}]`, `'${timestamp}'`, `${SITE_URL}`, `/${noiseWW}`],
  };

  try {
    await replaceInFile(replaceOptions);

    logger.info('[clean-scripts]: Workers updated.');
    process.exit(0);
  } catch (error) {
    logger.error(`[clean-scripts]: ${error}`);
    process.exit(1);
  }
})();

File: scripts/clean-scripts.go
/// 2>/dev/null ; gorun "$0" "$@" ; exit $?

package main

import (
	"log"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/samber/lo"
)

func init() {
	var files []string
	// runtime variables
	site := "https://cleverlaziness.xyz"
	now := time.Now()
	unix := now.Unix()

	// create src directory
	wd, err := os.Getwd()
	if err != nil {
		log.Fatal("[io.Getwd]:", err)
	}
	dist := filepath.Join(wd, "dist/")

	// create filter patterns
	typeIgnore := regexp.MustCompile(`.*(noise|sw).*\.js$`)
	typeMatch := regexp.MustCompile(`.*\.(js|css|woff|woff2)$`)

	// get files
	filepath.Walk(dist, func(path string, info os.FileInfo, err error) error {
		if err != nil {
			log.Fatal("[filepath.Glob]:", err)
		}
		// filter files
		ignore := typeIgnore.MatchString(path)
		matches := typeMatch.MatchString(path)

		// only save matches
		if !info.IsDir() && matches && !ignore {
			files = append(files, path)
		}

		return nil
	})
	files = lo.Map(files, func(item string, _ int) string {
		clean := strings.Replace(item, dist, "", -1)
		quotes := `"` + clean + `"`

		return quotes
	})

	// get service worker
	swRaw, err := os.ReadFile(filepath.Join(dist, "sw.js"))
	if err != nil {
		log.Fatal("[os.ReadFile]:", err)
	}
	sw := string(swRaw)

	// replace content
	assets := regexp.MustCompile(`\["staticAssets"\]`)
	baseURL := regexp.MustCompile(`baseURL`)
	version := regexp.MustCompile(`"version"`)
	swAssets := assets.ReplaceAllString(sw, "["+strings.Join(files, ",")+"]")
	swSite := baseURL.ReplaceAllString(swAssets, site)
	fmtSW := version.ReplaceAllString(swSite, `"`+strconv.FormatInt(unix, 10)+`"`)

	// replace sw.js
	newSW, err := os.Create(filepath.Join(dist, "sw.js"))
	if err != nil {
		log.Fatal("[os.Create]:", err)
	}
	defer newSW.Close()

	newSW.WriteString(fmtSW)
}
